### ISSUE TITLE
Integrate with ZeroNet?

### ISSUE THREAD

**Issue by leoloso** [leoloso/PoP#9]
Read discussion below.

_Note: previously this issue was about integration with React. But since the conversation on ZeroNet took over, I just renamed this issue._

**Comment by Snarkly** [comment]
Could the wordpress blog be integrated into another decentralized project named ZeroNet? Here's ZeroNet's github page. https://github.com/HelloZeroNet/ZeroNet Zeronet already has a blog but it has no CMS. Your wordpress looks a lot better than the ZeroNet blog.

**Comment by leoloso** [comment]
Yes, I think it can be done. Even though it is not possible to have WordPress running on ZeroNet (check [here](https://www.reddit.com/r/zeronet/comments/3ny2wo/can_i_host_a_wordpress_with_database_in_zeronet/) and [here](https://www.reddit.com/r/zeronet/comments/4vwpqb/zeronet_and_wordpress/)), I think that with some workaround it could work.

What I have in mind is to do something similar to issue #40 ([Make JSON requests immutable, access them through a CDN](https://github.com/leoloso/PoP/issues/40)), but replacing the CDN with ZeroNet. The key is that requests in PoP can be split into: dynamic (or mutable) and static (or immutable). Everything dynamic (eg: Recommend a post, log-in, create a post or add a comment) needs the server, like any traditional WordPress application. However, for reading content on the website, PoP produces the response in JSON code, and this code can be saved to disk and shared in a ZeroNet application. (For the CDN, as explained in #40, there is the extra step of adding a thumbprint to the file, but I'm not sure this is needed for ZeroNet)

Then, when first accessed, if the ZeroNet application is not hosting that static content, it needs request it from the server, after which it will both return it to the user, and share it among its peers in the BitTorrent network. Static content is generated on demand, instead of all of it at once, since the combination of all pages to be generated is huge. If the user is offline, he/she won't be able to interact with the WordPress application, but will be able to access all already-generated static content from ZeroNet.

In summary, the ZeroNet application will be able to show all static content, however it will need to fall back on the server for all dynamic content and user interactions. It is not a pure ZeroNet application, but a hybrid with a traditional WordPress app.

**Comment by Snarkly** [comment]
Thanks for explaining it, now I just need to find someone that can do it. I'm not a coder so I wouldn't be able to do it.

**Comment by fiskr** [comment]
I would like to help. I am a "coder". 😄

**Comment by leoloso** [comment]
@fiskr thanks for offering to help, that's awesome :)

Yes, I agree we need to move it out from this issue, however I wonder if it actually belongs to this repository. I see three possibilities where to post it:

1. Issue under PoP
2. Issue under ZeroNet
3. Issue under a new repository: integration of PoP for ZeroNet

I think the 3rd option is the best. What do you think? Then, you could be the owner of that repository since you'll take care of the ZeroNet implementation, and start work on it from the ZeroNet side. 

On my side, I'll work on issue #40 which will integrate PoP with a CDN (I can have it ready in 3 weeks I think, I'm currently busy with another issue), and then we work together on how to replace the CDN bit from PoP and hook in ZeroNet instead.

What do you think?

**Comment by fiskr** [comment]
> Then, you could be the owner of that repository since you'll take care of the ZeroNet implementation, and start work on it from the ZeroNet side.

I have zero experience with ZeroNet or PoP - I got here because I noticed the community is really driving for integration with a CMS (and WordPress being a popular CMS, is widely mentioned). Someone linked this issue from ZeroTalk, and brought me to this issue and here we are. I have lots of experience with CSS and javascript, and experience in Java, *nix, and other back-end technologies (my job is basically in web and supporting websites with [AEM](https://www.adobe.com/marketing-cloud/enterprise-content-management.html)).

I would love to help, even owning the repository - but I will need to take a lot of time to find people to help me learn PoP, CDN, and ZeroNet to even understand what a potential implementation would be. My first attempt will also probably be garbage, so I will need lots of advice and peer review along the way. Not sure what the resources are, or who could help with that.

Maybe we should reach out to someone on the ZeroNet side of things to see how they feel about the approach we should take (i.e. whether to put it on the PoP side, ZeroNet side, or in a new repo).

**Comment by fiskr** [comment]
I have looked around, it looks like https://github.com/HelloZeroNet/ZeroAdmin was created from [this issue](https://github.com/HelloZeroNet/ZeroNet/issues/354) and may be heavily related. They want a zite (ZeroNet site) that acts as an admin for modifying the ZeroBlogs (including I assume WordPress features, like modifying content, theme, etc.)...

**Comment by leoloso** [comment]
I would be happy to host it under ZeroAdmin. Let's wait and see what they say then.

About your involvement in the potential ZeroNet + PoP integration, I can certainly teach you about PoP, however since I'll implement #40 myself, I believe the PoP side will be pretty much already covered, and quite likely you won't need to work on it... from the ZeroNet side I don't know, I guess you would need to find out with them how to work it out. About the design on how the 2 systems will interact with each other, we can work on that together, I have some ideas already.

**Comment by fiskr** [comment]
Well, for example, to get WordPress working (on something like a ZeroAdmin zite), what would I need to do? I know PoP abstracts out the data requests WordPress needs - so do I just need to setup generic API calls for PoP? How does PoP work, and what is involved in implementing it?

Maybe we can look at other examples of people implementing PoP for other sites...

**Comment by leoloso** [comment]
Actually, the integration between ZeroNet and PoP as I suggested is quite detached. Not only ZeroNet does not need to call PoP's API, but in a way, it won't even care about PoP. My suggestion is that ZeroNet acts as a CDN: when requesting a URL, if it has that response, it serves it; if not, it goes fetch it from some upstream server, gets the response, "caches" it (i.e. distributes it in the Tor network) and serves it. ZeroNet is just requesting a URL, so there's no API called or anything, but it's a standard HTTP request.

The beauty about this design is that it is not tied to PoP or even WordPress; you can hook any application as the upstream server. I recommended PoP, but it could perfectly be something else too. Why PoP can work as an upstream server? Because it can generate static files, in JSON format, for the dynamic content, and these static files can be distributed in the Tor network. If in the future you want to switch to, let's say, a static site generator, you could just unplug PoP and plug in the new system, and it should keep working perfectly.

This approach could be considered a drawback since this is not a pure ZeroNet application, but a hybrid with a traditional application. But it can also be considered an advantage, since it opens the doors to traditional applications to distribute their content through ZeroNet. I think that's quite cool!

**Comment by leoloso** [comment]
I changed the title of this issue, so now it's about ZeroNet

**Comment by leoloso** [comment]
So, this hybrid approach will need:

- A traditional website, under http://www.website.com
- A ZeroNet website, to distribute the content from the website, under http://localhost:43110/13DNDkMUExRf9Xa9ogwPKqp7zyHFEqbhC2

The traditional website is a PoP website, which can change all its links and sources to the ZeroNet website, so every single request is first routed through ZeroNet.

The user initially goes to http://localhost:43110/13DNDkMUExRf9Xa9ogwPKqp7zyHFEqbhC2. All assets, such as images, are served from this domain. So, if the website wants to request image "img/logo.jpg", it will:

- Request http://localhost:43110/13DNDkMUExRf9Xa9ogwPKqp7zyHFEqbhC2/img/logo.jpg
- If the image is there, serve it
- If it is not, the ZeroNet website will fetch it from http://www.website.com/img/logo.jpg
- When it gets it, it will share it on the Tor network, and then serve it

Next time anyone requests that image, it will be already on the ZeroNet Tor network.

Same will happen with content. If anyone clicks on the "posts" link, the PoP website will:

- Request http://localhost:43110/13DNDkMUExRf9Xa9ogwPKqp7zyHFEqbhC2/posts/?output=json, which outputs the content of the page in JSON format
- If it's there, serve it
- If not, same as above

What happens if there is a new post, so the content in /posts changes? How will ZeroNet know? Because it can constantly ping the traditional website for a "post thumbprint", representing the last time content was added, and access the resources using that thumbprint, eg:

- http://localhost:43110/13DNDkMUExRf9Xa9ogwPKqp7zyHFEqbhC2/posts/?output=json&thumbprint=28497726373822229

If ZeroNet doesn't have that resource, it will:

- Request http://www.website.com/posts/?output=json&thumbprint=28497726373822229
- Get the response, cache it (distribute it in Tor) under both 
 - http://localhost:43110/13DNDkMUExRf9Xa9ogwPKqp7zyHFEqbhC2/posts/?output=json&thumbprint=28497726373822229
 http://localhost:43110/13DNDkMUExRf9Xa9ogwPKqp7zyHFEqbhC2/posts/?output=json
- Return it to the user

If there is no connection to the internet, or the website is down, ZeroNet can then return the previous response (possibly stale) for that page, which is that URL without the thumbprint.

Please notice that ZeroNet is accessed only for GET requests. For POST, there is no need to route it through ZeroNet, it will go straight to the website. If the website is down, then you can't post content (or log in, or add a comment), but you can still access all already-generated content from ZeroNet

**Comment by fiskr** [comment]
Hmm, I think this is problematic - ZeroNet shouldn't have dependencies on the a normal website. This feels like a workaround, when ideally you would just integrate the WordPress admin site into a zite that produces other zites (as opposed to websites on the clearnet). 

Your approach would work, but against the philosophy and design of ZeroNet, and I worry about that.

**Comment by leoloso** [comment]
@fiskr I'll start implementing #40 from today, it will take possibly one week, I think after that we can discuss it again since the picture will be more clear, it will be possible to visualize my concept. We are also still waiting for @HelloZeroNet to say if they'd be interested...

In the meantime, would you be interested in contributing to PoP? This is the list of issues I need help with:

https://github.com/leoloso/PoP/labels/help%20wanted

**Comment by fiskr** [comment]
> In the meantime, would you be interested in contributing to PoP? This is the list of issues I need help with:
> https://github.com/leoloso/PoP/labels/help%20wanted

I would love to help out! I will need a lot of help getting a local development environment setup, and getting accustomed to everything, but we can discuss the issues as they come up. Let me find an issue that looks like I could easily start with and go from there.

**Comment by leoloso** [comment]
I'll be happy if you can start with #38 since it's more of a WordPress stuff than PoP, so it's easier to start with. I'll switch to that issue to think on how to tackle it. 

Do you have experience with WordPress? How much? You know how to develop a plug-in? hooks? If you do, everything is set, we can start with that issue. If not, you will need to play with some local WordPress installation first.

**Comment by fiskr** [comment]
I think I will need to play with a local installation first. I have almost no experience with WordPress, especially from a development perspective.

**Comment by leoloso** [comment]
@fiskr you said you like scripting. Do you have experience working with grunt/gulp? We should make it easier to run the scripts to optimize the .css and .js files in PoP (i.e. minifying + mangling), right now it's set-up calling Google's min and UglifyJS in an ugly .sh file, setting up plenty of environment variables... it should be better than that, more automated, easier to execute.

Would you like to take care of this task?

**Comment by fiskr** [comment]
Yeah - it's just a matter of finding the time, and getting familiar with PoP and the existing code.

**Comment by leoloso** [comment]
Hi @fiskr sorry for the late response (it was a busy week)

Yes, that is the .sh file I mentioned. You can see how there are environment variables declared to minify the files... having to set-up variables in the system for the scripts to work is certainly horrible and it doesn't scale up. So something more proper is needed.

I just created the issue here: #49. Please be welcome to help :)

**Comment by krzysztof113** [comment]
Is there a Zeronet site with POP to clone or working one?

**Comment by leoloso** [comment]
Nops

**Comment by leoloso** [comment]
This will never happen
